File: sdks/java/reactor-sdk/src/test/java/inc/reactor/sdk/internal/CompletionsTest.java
package inc.reactor.sdk.internal;

import static java.lang.foreign.ValueLayout.JAVA_INT;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import inc.reactor.sdk.AbortedException;
import inc.reactor.sdk.DecodeFailedException;
import inc.reactor.sdk.RateLimitedException;
import inc.reactor.sdk.ReactorException;
import java.lang.foreign.Arena;
import java.lang.foreign.FunctionDescriptor;
import java.lang.foreign.Linker;
import java.lang.foreign.MemorySegment;
import java.lang.invoke.MethodHandle;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * The completion bridge, driven through a real function pointer rather than by calling the
 * dispatch method directly — a test that calls the helper proves the helper works.
 */
final class CompletionsTest {

    @Test
    @DisplayName("a successful completion decodes, then settles")
    void successDecodesThenSettles() throws Exception {
        try (Arena arena = Arena.ofShared()) {
            Completions completions = new Completions(arena);
            CompletableFuture<String> future = new CompletableFuture<>();
            Completions.Ticket ticket = completions.register("get_stats", json -> json + "!", future);

            fire(arena, ticket, 1, "{\"a\":1}", null);

            assertEquals("{\"a\":1}!", future.get());
            assertEquals(0, completions.pendingCount(), "a settled operation must not stay registered");
        }
    }

    @Test
    @DisplayName("a successful payload that will not decode is a decode failure, not an empty answer")
    void aPayloadThatWillNotDecodeIsADecodeFailure() {
        try (Arena arena = Arena.ofShared()) {
            Completions completions = new Completions(arena);
            CompletableFuture<String> future = new CompletableFuture<>();
            Completions.Ticket ticket = completions.register(
                    "request_schema",
                    json -> {
                        throw new IllegalArgumentException("no 'commands' field");
                    },
                    future);

            fire(arena, ticket, 1, "{}", null);

            ExecutionException thrown = assertThrows(ExecutionException.class, future::get);
            DecodeFailedException cause = assertInstanceOf(DecodeFailedException.class, thrown.getCause());
            assertEquals("DECODE_FAILED", cause.code());
            assertTrue(cause.getMessage().contains("request_schema"), cause.getMessage());
            // The point of the whole rule: the caller does not get an empty schema they cannot
            // tell apart from a model that declares nothing.
            assertFalse(future.isCompletedExceptionally() && future.isCancelled());
        }
    }

    @Test
    @DisplayName("a failed completion becomes the typed exception its code names")
    void failureBecomesATypedException() {
        try (Arena arena = Arena.ofShared()) {
            Completions completions = new Completions(arena);
            CompletableFuture<String> future = new CompletableFuture<>();
            Completions.Ticket ticket = completions.register("connect", json -> json, future);

            fire(
                    arena,
                    ticket,
                    0,
                    null,
                    "{\"code\":\"RATE_LIMITED\",\"message\":\"slow down\",\"recoverable\":true,"
                            + "\"status\":429,\"operation\":\"connect\",\"retry_after_ms\":2500}");

            ExecutionException thrown = assertThrows(ExecutionException.class, future::get);
            RateLimitedException cause = assertInstanceOf(RateLimitedException.class, thrown.getCause());
            assertEquals(429, cause.status());
            assertEquals("connect", cause.operation());
            assertEquals(2500L, cause.retryAfterMs());
            assertTrue(cause.isRecoverable());
        }
    }

    @Test
    @DisplayName("a completion arriving after close finds nothing, and settles nothing twice")
    void aLateCompletionFindsNothing() {
        try (Arena arena = Arena.ofShared()) {
            Completions completions = new Completions(arena);
            CompletableFuture<String> future = new CompletableFuture<>();
            Completions.Ticket ticket = completions.register("send_command", json -> json, future);

            completions.settleAll(DetachedCompletion.abandoned("send_command"));
            ExecutionException thrown = assertThrows(ExecutionException.class, future::get);
            assertInstanceOf(AbortedException.class, thrown.getCause());

            // The FFI does not know the client closed. This is the callback arriving anyway.
            fire(arena, ticket, 1, "{}", null);

            assertInstanceOf(
                    AbortedException.class,
                    assertThrows(ExecutionException.class, future::get).getCause(),
                    "the late completion must not overwrite what the caller was already told");
        }
    }

    @Test
    @DisplayName("close settles every pending operation, rather than leaving callers waiting forever")
    void closeSettlesEverythingPending() {
        try (Arena arena = Arena.ofShared()) {
            Completions completions = new Completions(arena);
            CompletableFuture<String> one = new CompletableFuture<>();
            CompletableFuture<String> two = new CompletableFuture<>();
            completions.register("connect", json -> json, one);
            completions.register("get_stats", json -> json, two);
            assertEquals(2, completions.pendingCount());

            completions.settleAll(DetachedCompletion.abandoned("close"));

            assertTrue(one.isCompletedExceptionally());
            assertTrue(two.isCompletedExceptionally());
            assertEquals(0, completions.pendingCount());
        }
    }

    @Test
    @DisplayName("an error payload that is not JSON still becomes an error, not a lost failure")
    void anUnparseableErrorIsStillAnError() {
        try (Arena arena = Arena.ofShared()) {
            Completions completions = new Completions(arena);
            CompletableFuture<String> future = new CompletableFuture<>();
            Completions.Ticket ticket = completions.register("connect", json -> json, future);

            fire(arena, ticket, 0, null, "this is not JSON");

            ReactorException cause = assertInstanceOf(
                    ReactorException.class,
                    assertThrows(ExecutionException.class, future::get).getCause());
            assertEquals("DECODE_FAILED", cause.code());
            assertTrue(cause.getMessage().contains("this is not JSON"), cause.getMessage());
        }
    }

    /** Calls the completion callback the way the FFI would: through its function pointer. */
    @SuppressWarnings("restricted") // downcallHandle: calling our own stub as C would
    private static void fire(Arena arena, Completions.Ticket ticket, int ok, String resultJson, String errorJson) {
        MethodHandle call = Linker.nativeLinker()
                .downcallHandle(
                        ticket.callback(),
                        FunctionDescriptor.ofVoid(
                                JAVA_INT,
                                java.lang.foreign.ValueLayout.ADDRESS,
                                java.lang.foreign.ValueLayout.ADDRESS,
                                java.lang.foreign.ValueLayout.ADDRESS));
        MemorySegment result = resultJson == null ? MemorySegment.NULL : arena.allocateFrom(resultJson);
        MemorySegment error = errorJson == null ? MemorySegment.NULL : arena.allocateFrom(errorJson);
        try {
            call.invokeExact(ok, result, error, ticket.userdata());
        } catch (Throwable t) {
            throw new AssertionError("the completion callback threw out of the stub", t);
        }
    }
}

File: sdks/android/reactor-sdk-android/src/test/kotlin/inc/reactor/sdk/android/CompletionsTest.kt
package inc.reactor.sdk.android

import inc.reactor.sdk.android.internal.Completions
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Deferred
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.async
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertThrows
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * The completion registry: settle once, decode before claiming, and never leave a caller waiting.
 *
 * The failing paths capture their exception with `runCatching` *inside* the coroutine rather than
 * letting `await()` rethrow it. Under structured concurrency a failing `async` cancels its parent
 * — here `runTest`'s own scope — so catching at the await site is too late and the whole test
 * fails with the right exception for the wrong reason. This suite got that wrong first.
 */
@OptIn(ExperimentalCoroutinesApi::class)
class CompletionsTest {
    @Test
    fun `a successful completion resolves with the decoded value`() =
        runTest {
            val ticket = CompletableDeferred<Long>()
            val result =
                async {
                    Completions.await("connect", decode = { it ?: "none" }) { ticket.complete(it) }
                }
            Completions.settle(ticket.await(), ok = true, resultJson = "{}", errorJson = null)
            assertEquals("{}", result.await())
        }

    @Test
    fun `a failed completion throws the typed error`() =
        runTest {
            val ticket = CompletableDeferred<Long>()
            val result =
                async {
                    runCatching { Completions.await("send_command", decode = { it }) { ticket.complete(it) } }
                }
            Completions.settle(
                ticket.await(),
                ok = false,
                resultJson = null,
                errorJson = """{"code":"RATE_LIMITED","message":"slow down","retry_after_ms":250}""",
            )
            val error = awaitFailure<RateLimitedException>(result)
            assertEquals(250L, error.retryAfterMs)
            assertTrue(error.recoverable)
        }

    /**
     * The failure the ordering exists to prevent: if the promise were claimed first and the
     * payload converted after, this throw would land where nothing can be settled any more and
     * the caller would hang instead of seeing DECODE_FAILED.
     */
    @Test
    fun `a success whose payload will not decode fails with DECODE_FAILED`() =
        runTest {
            val ticket = CompletableDeferred<Long>()
            val result =
                async {
                    runCatching {
                        Completions.await<String>("request_schema", decode = { error("bad shape") }) {
                            ticket.complete(it)
                        }
                    }
                }
            Completions.settle(ticket.await(), ok = true, resultJson = """{"unexpected":1}""", errorJson = null)
            val error = awaitFailure<DecodeFailedException>(result)
            assertTrue(
                "the raw payload belongs in the message — it is what a decode bug is diagnosed from",
                error.message!!.contains("unexpected"),
            )
        }

    /** An absent payload is an answer: a void operation reports nothing and that is success. */
    @Test
    fun `an absent payload is distinct from a malformed one`() =
        runTest {
            val ticket = CompletableDeferred<Long>()
            val result =
                async {
                    Completions.await("disconnect", decode = { it == null }) { ticket.complete(it) }
                }
            Completions.settle(ticket.await(), ok = true, resultJson = null, errorJson = null)
            assertEquals(true, result.await())
        }

    @Test
    fun `settling twice is ignored`() =
        runTest {
            val ticket = CompletableDeferred<Long>()
            val result =
                async {
                    Completions.await("connect", decode = { it ?: "first" }) { ticket.complete(it) }
                }
            val id = ticket.await()
            Completions.settle(id, ok = true, resultJson = "first", errorJson = null)
            // A second completion for the same ticket must not throw, and must not change the answer.
            Completions.settle(id, ok = false, resultJson = null, errorJson = """{"code":"CONFLICT"}""")
            assertEquals("first", result.await())
        }

    @Test
    fun `an unknown ticket is dropped rather than throwing`() {
        Completions.settle(ticket = 999_999L, ok = true, resultJson = "{}", errorJson = null)
    }

    /** Teardown must settle waiters: an unresolved promise waits for the life of the process. */
    @Test
    fun `abandonAll settles outstanding callers`() =
        runTest {
            val ticket = CompletableDeferred<Long>()
            val result =
                async {
                    runCatching { Completions.await("download_clip", decode = { it }) { ticket.complete(it) } }
                }
            ticket.await()
            Completions.abandonAll("the client was closed; the download may still be writing")
            val error = awaitFailure<AbortedException>(result)
            assertTrue(
                "the message should say the native work may continue, not claim it stopped",
                error.message!!.contains("may still be writing"),
            )
        }

    @Test
    fun `a failure to start removes the ticket`() =
        runTest {
            val before = Completions.pendingCount
            val result =
                async {
                    runCatching {
                        Completions.await<String>("connect", decode = { it!! }) {
                            error("the native call refused")
                        }
                    }
                }
            awaitFailure<IllegalStateException>(result)
            assertEquals(
                "a completion that never started must not stay pending forever",
                before,
                Completions.pendingCount,
            )
        }

    // ── timeoutMillis ────────────────────────────────────────────────────────

    @Test
    fun `NaN is a caller bug rather than a duration`() {
        val error =
            assertThrows(BadRequestException::class.java) {
                Completions.timeoutMillis(Double.NaN)
            }
        assertEquals("BAD_REQUEST", error.code)
    }

    @Test
    fun `infinite and negative both mean no bound`() {
        assertEquals(0L, Completions.timeoutMillis(Double.POSITIVE_INFINITY))
        assertEquals(0L, Completions.timeoutMillis(Double.NEGATIVE_INFINITY))
        assertEquals(0L, Completions.timeoutMillis(-1.0))
        assertEquals(0L, Completions.timeoutMillis(null))
    }

    @Test
    fun `an enormous timeout saturates rather than wrapping`() {
        assertEquals(Long.MAX_VALUE, Completions.timeoutMillis(Double.MAX_VALUE))
        assertEquals(1500L, Completions.timeoutMillis(1.5))
    }

    /**
     * Await a [Deferred] expecting it to fail with [E].
     *
     * Written out rather than wrapping `assertThrows` around a `runBlocking`: nesting a blocking
     * builder inside `runTest`'s scheduler deadlocks, which is how the first version of this
     * suite failed. Awaiting in the test's own coroutine is both correct and the thing being
     * tested — the caller really does suspend here.
     */
    private suspend inline fun <reified E : Throwable> awaitFailure(deferred: Deferred<Result<*>>): E {
        val thrown = deferred.await().exceptionOrNull()
        assertTrue(
            "expected ${E::class.simpleName}, got ${thrown?.let { it::class.simpleName } ?: "success"}",
            thrown is E,
        )
        return thrown as E
    }
}
